### PR TITLE
Refactoring of MethodBinder.Bind

### PR DESCRIPTION
**Pull request by lostmsu**
### What does this implement/fix? Explain your changes.
This is a refactoring/cleaning up of MethodBinder.Bind method. No new features were introduced.

### Does this close any currently open issues?
This is a preliminary work necessary for #823.

### Any other comments?
The refactoring was mostly done using ReSharper with little manual interventions. The semantics should be preserved.

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [x] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by lostmsu** [comment]
The build failure seems to be simple build timeout. Very unlikely caused by this change.

**Comment by filmor** [comment]
I'll have a closer look at this at the weekend. No idea why travis is not getting triggered here, I've poked the one part that timed out though, so this one should go green.

Thanks already :)

**Comment by filmor** [comment]
I'll just close and reopen this one to poke Travis...

**Comment by lostmsu** [comment]
CI still having problems :/

File: src/runtime/methodbinder.cs [hunk 1]
```diff
@@ -301,192 +299,242 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
...
+
+                object target = null;
+                if (!mi.IsStatic && inst != IntPtr.Zero)
                 {
-                    match = true;
-                    defaultArgList = new ArrayList();
-                    for (var v = pynargs; v < clrnargs; v++)
+                    //CLRObject co = (CLRObject)ManagedType.GetManagedObject(inst);
+                    // InvalidCastException: Unable to cast object of type
+                    // 'Python.Runtime.ClassObject' to type 'Python.Runtime.CLRObject'
```
Comment (filmor) [inline src/runtime/methodbinder.cs]: This comment confuses me more than it helps ;). It's clear after reading the next one, but here not so much.

File: src/runtime/methodbinder.cs [hunk 2]
```diff
@@ -301,192 +299,242 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
...
+                if (!TryConvertArgument(op, parameter.ParameterType, needsResolution, out margs[paramIndex], out isOut))
+                {
+                    return null;
+                }
+
+                if (arrayStart == paramIndex)
+                {
+                    // TODO: is this a bug? Should this happen even if the conversion fails?
+                    // GetSlice() creates a new reference but GetItem()
+                    // returns only a borrow reference.
```
Comment (filmor) [inline src/runtime/methodbinder.cs]: This indeed looks like a bug to me, the code has to also run before `return null`.

**Comment by filmor** [comment]
I think both comments are for things that you didn't change, just wanted to leave them around.

**Pull request merged**